### PR TITLE
HDDS-8374. Fix Flaky Unit Test for TestContainerStateCounts in Recon

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

The testDeletedContainerCount() method of TestContainerStateCounts is considered flaky due to the processing order of the container report. If the pipeline does not exist, then the container is not added until the next container report is processed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8374

## How was this patch tested?

Ran the UT several times to make sure it's not flaky

### CODE REVIEW

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestContainerStateCounts.java [hunk 1]
```diff
@@ -387,4 +397,20 @@ public void testDeletedContainerCount() throws Exception {
     Assertions.assertEquals(3, clusterStateResponse1.getOpenContainers());
   }
 
+  private void waitAndCheckConditionAfterHeartbeat(Callable<Boolean> check)
+      throws Exception {
+    // if container report is processed first, and pipeline does not exist
+    // then container is not added until the next container report is processed
+    StorageContainerDatanodeProtocolProtos.SCMHeartbeatRequestProto
+        heartbeatRequestProto =
+        StorageContainerDatanodeProtocolProtos.SCMHeartbeatRequestProto.newBuilder()
```
Comment (adoroszlai) [inline hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestContainerStateCounts.java:406]:
```
hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestContainerStateCounts.java
 406: Line is longer than 80 characters (found 84).
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @ArafatKhan2198 for the fix.  I'll try repeated runs.  In the meantime, please fix checkstyle error.

**Comment by adoroszlai** [comment]
@ArafatKhan2198 seems like the test failed in the first run:

```
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 37.517 s <<< FAILURE! - in org.apache.hadoop.ozone.recon.api.TestContainerStateCounts
Error:  org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.testContainerStatsCount  Time elapsed: 37.468 s  <<< ERROR!
java.util.concurrent.TimeoutException: timeout: after 30000 millis
	at org.apache.ozone.test.LambdaTestUtils$GenerateTimeout.evaluate(LambdaTestUtils.java:677)
	at org.apache.ozone.test.LambdaTestUtils.await(LambdaTestUtils.java:154)
	at org.apache.ozone.test.LambdaTestUtils.await(LambdaTestUtils.java:192)
	at org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.waitAndCheckConditionAfterHeartbeat(TestContainerStateCounts.java:413)
	at org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.testContainerStatsCount(TestContainerStateCounts.java:369)
```

https://github.com/ArafatKhan2198/ozone/actions/runs/4608724963/jobs/8144896789#step:6:2706

**Comment by ArafatKhan2198** [comment]
> @ArafatKhan2198 seems like the test failed in the first run:
> 
> ```
> Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 37.517 s <<< FAILURE! - in org.apache.hadoop.ozone.recon.api.TestContainerStateCounts
> Error:  org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.testContainerStatsCount  Time elapsed: 37.468 s  <<< ERROR!
> java.util.concurrent.TimeoutException: timeout: after 30000 millis
> 	at org.apache.ozone.test.LambdaTestUtils$GenerateTimeout.evaluate(LambdaTestUtils.java:677)
> 	at org.apache.ozone.test.LambdaTestUtils.await(LambdaTestUtils.java:154)
> 	at org.apache.ozone.test.LambdaTestUtils.await(LambdaTestUtils.java:192)
> 	at org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.waitAndCheckConditionAfterHeartbeat(TestContainerStateCounts.java:413)
> 	at org.apache.hadoop.ozone.recon.api.TestContainerStateCounts.testContainerStatsCount(TestContainerStateCounts.java:369)
> ```
> 
> https://github.com/ArafatKhan2198/ozone/actions/runs/4608724963/jobs/8144896789#step:6:2706

My apologise @adoroszlai. By mistake, I reverted back to the previous value after running the tests successfully. Please run the tests now. I have already run them a total of 20 times, and they have passed.

image

**Comment by adoroszlai** [comment]
Thanks @ArafatKhan2198 for updating the patch.  Have you run repeated tests in CI?  Can you please share the link to the CI run?

**Comment by ArafatKhan2198** [comment]
> https://github.com/adoroszlai/hadoop-ozone/actions/runs/4610115793

Here you go @adoroszlai  :- https://github.com/ArafatKhan2198/ozone/actions/runs/4669283289

**Comment by adoroszlai** [comment]
Thanks @ArafatKhan2198 for updating the patch and for repeated runs, the results look good.  I'll defer review of the actual fix to folks more familiar with Recon.

**Comment by ArafatKhan2198** [comment]
@devmadhuu @adoroszlai Can you please take a look

**Review by smengcl — APPROVED** [review]
Thanks @ArafatKhan2198 . lgtm

**Comment by adoroszlai** [comment]
Please try to avoid force-push when updating the PR.  Here are some great articles that explain why:

https://developers.mattermost.com/blog/submitting-great-prs/#4-avoid-force-pushing
https://www.freecodecamp.org/news/optimize-pull-requests-for-reviewer-happiness#request-a-review

**Comment by adoroszlai** [comment]
> Therefore, I have decided to incorporate this small name change in this PR.

It would be better to avoid incorporating unrelated changes.

1. We already had successful CI for this PR, could have been merged.  Now we need to trigger another run.
2. Commit message would be confusing (fix flaky test vs. rename JSON element).
3. CI needs to run fewer checks if only unit test is changed.

**Comment by ArafatKhan2198** [comment]
> > Therefore, I have decided to incorporate this small name change in this PR.
> 
> It would be better to avoid incorporating unrelated changes.
> 
> 1. We already had successful CI for this PR, could have been merged.  Now we need to trigger another run.
> 2. Commit message would be confusing (fix flaky test vs. rename JSON element).
> 3. CI needs to run fewer checks if only unit test is changed.

Understood, I have removed those changes. 
Thanks for the feedback

**Pull request merged**